x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/sender/HttpRequestExecutorService.java

```diff
@@ -0,0 +1,177 @@
+ * requests until connections are available.
+ *
+ * <b>NOTE:</b> It is the responsibility of the class constructing the
+ * {@link org.apache.http.client.methods.HttpUriRequest} to set a timeout for how long this executor will wait
+ * attempting to execute a task (aka waiting for the connection manager to lease a connection). See
+ * {@link org.apache.http.client.config.RequestConfig.Builder#setConnectionRequestTimeout} for more info.
+ */
+class HttpRequestExecutorService extends AbstractExecutorService {
+    private static final Logger logger = LogManager.getLogger(HttpRequestExecutorService.class);
+
+    private final ThreadContext contextHolder;
+    private final String serviceName;
+    private final BlockingQueue<HttpTask> queue;
+    private final AtomicBoolean running = new AtomicBoolean(true);
+    private final CountDownLatch terminationLatch = new CountDownLatch(1);
+    private final HttpClientContext httpContext;
+    private final HttpClient httpClient;
+
+    @SuppressForbidden(reason = "properly rethrowing errors, see EsExecutors.rethrowErrors")
+    HttpRequestExecutorService(ThreadContext contextHolder, String serviceName, HttpClient httpClient) {
+        this(contextHolder, serviceName, httpClient, null);
+    }
+
+    @SuppressForbidden(reason = "properly rethrowing errors, see EsExecutors.rethrowErrors")
+    HttpRequestExecutorService(ThreadContext contextHolder, String serviceName, HttpClient httpClient, @Nullable Integer capacity) {
+        this.contextHolder = Objects.requireNonNull(contextHolder);
+        this.serviceName = Objects.requireNonNull(serviceName);
+        this.httpClient = Objects.requireNonNull(httpClient);
+        this.httpContext = HttpClientContext.create();
+
+        if (capacity == null) {
+            this.queue = new LinkedBlockingQueue<>();
+        } else {
+            this.queue = new LinkedBlockingQueue<>(capacity);
+        }
+    }
+
+    /**
+     * Begin servicing tasks.
+     */
+    public void start() {
+        try {
+            while (running.get()) {
+                HttpTask task = queue.take();
+                if (task.shouldShutdown() || running.get() == false) {
+                    running.set(false);
+                    logger.debug(() -> format("Http executor service [%s] exiting", serviceName));
+                } else {
+                    executeTask(task);
+                }
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        } finally {
+            terminationLatch.countDown();
+        }
+    }
+
+    private void executeTask(HttpTask task) {
+        try {
+            contextHolder.preserveContext(task).run();
+        } catch (Exception e) {
+            logger.error(format("Http executor service [%s] failed to execute request [%s]", serviceName, task), e);
+        }
+        EsExecutors.rethrowErrors(ThreadContext.unwrap(task));
```

Comment:
`rethrowErrors` doesn't actually throw it returns a non-null value if there is something to throw. 

I see that `AbstractProcessWorkerExecutorService` has made the same mistake: https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/AbstractProcessWorkerExecutorService.java#L115

The only proper usage I can find is in Scheduler.java

https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/threadpool/Scheduler.java#L285

`EsExecutors.rethrowErrors` only applies to `RunnableFuture` i.e. runnables that return a value, the HttpTasks are `Runnables` they don't return return a value as the listener is called. I think as long as the `start()` handles the `InterruptedException` this isn't necessary and this line can be removed.

The confusing usage has been inherited from `AbstractProcessWorkerExecutorService` where the mistake was originally made could you remove the call from there and here please.

Reply:
Ah good point. Should we remove the `ThreadContext contextHolder` too? I think that's only used for the `rethrowErrors`.

> could you remove the call from there and here please.

Yeah definitely. I'll do the changes in `AbstractProcessWorkerExecutorService` in a separate PR if that's ok.

Reply:
Yeah I think you're right, I saw the failing test.